include/net/nrf_cloud_coap.h

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef NRF_CLOUD_COAP_H_
+#define NRF_CLOUD_COAP_H_
+
+/** @file nrf_cloud_coap.h
+ * @brief Module to provide nRF Cloud CoAP API
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <net/nrf_cloud_rest.h>
+#include <net/nrf_cloud_agps.h>
+#include <net/nrf_cloud_pgps.h>
+#include <zephyr/net/coap_client.h>
+
+/**
+ * @defgroup nrf_cloud_coap nRF CoAP API
+ * @{
+ */
+
+/* Transport functions */
+/**@brief Initialize nRF Cloud CoAP library.
+ *
+ * @return 0 if initialization was successful, otherwise, a negative error number.
+ */
+int nrf_cloud_coap_init(void);
+
+/**@brief Connect to nRF Cloud CoAP server.
+ *
+ * Upon successful connection, call @ref nrf_cloud_coap_is_authorized to confirm
+ * whether the device was granted access to nRF Cloud.
+ *
+ * @return 0 if connected successfully, otherwise, a negative error number.
+ */
+int nrf_cloud_coap_connect(void);
+
+/**@brief Check if device is authorized to use nRF Cloud CoAP.
+ *
+ * A device is authorized if the JWT it POSTed to the /auth/jwt endpoint is valid
+ * and matches credentials for a device already provisioned and associated with nRF Cloud.
+ *
+ * @retval true Device is allowed to access nRF Cloud services.
+ * @retval false Device is disallowed from accessing nRF Cloud services.
+ */
+bool nrf_cloud_coap_is_authorized(void);
```

Comment:
Our system is unique in that unlike a normal CoAP server and client, where if the DTLS handshake with mutual authentication fails that means the client is not connected as well as not authorized, in our system, only the client authenticates the server using the root CA cert.  It is then connected with a fully encrypted channel.  But, until the client does a `POST` to the `/auth/jwt` "end point" with a JWT that is for an authorized device, it cannot use any services.

Really, I suppose I should rename the function to be `nrf_cloud_coap_authenticate()`, since that is what it actually does.  The `nrf_cloud_coap_init()` function also calls connect() on the socket, but the way UDP/DTLS works is that connect() is kind of symbolic, and the DTLS handshake is deferred to the first attempted transfer from the client to the server.

So, between calling `nrf_cloud_coap_init()` and `nrf_cloud_coap_connect()`, any attempt to access a service will (inside the network stack) cause the DTLS handshake to occur, but the service will fail due to lack of authorization.  So the `nrf_cloud_coap_is_authorized()` function does serve a purpose.

Reply:
renamed again per other reviewers